<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Lei Li leil@uchicago.edu" />

<meta name="date" content="2020-07-02" />

<title>Librator, a platform for optimized sequence editing, design, and expression of influenza virus proteins</title>
<link rel="icon" type="image/x-icon" href="logo.ico"/>

<script src="https://ajax.aspnetcdn.com/ajax/jQuery/jquery-1.8.0.js"></script>
<link rel="stylesheet" type="text/css" href="css/style.css">


</head>

<body>

  <div class="navbox">
    <ul class="nav">
      <li><a class="active" href="index.html">About</a></li>
      <li><a href="install.html">Download&Install</a></li>
      <li><a href="tutorial.html">Tutorial</a></li>
      <li><a href="video_tutorial.html">Video Tutorial</a></li>
      <li><a href="contact.html">Contact</a></li>
      <li><a href="https://github.com/WilsonImmunologyLab/Librator">Github Page</a></li>
    </ul>
  </div>

<div class="row">
<!--
  <div class="left">
    <div class="center">
      <img class="logo" src="img/logo.png" style="margin:auto;">
      <span class="author_name">Librator, a platform for optimized sequence editing, design, and expression of influenza virus proteins</span>
      <span class="author_bio"><a href="http://wilsonlab.uchicago.edu/" target="blank">Wilson Lab</a></span>
      <span class="author_job">The University of Chicago </span>
    </div>
  </div>
-->
  <div class="col s12 m3">
    <div class="table cover">
      <div class="cover-card table-cell table-middle">
        <div class="center">
          <img class="logo" src="img/logo.png">
        </div>
        
        <span class="author_name">Librator, a platform for optimized sequence editing, design, and expression of influenza virus proteins</span>
        <span class="author_bio">Wilson Lab</span>
        <span class="author_job">The University of Chicago </span>
      </div>
    </div>
  </div>

  <div class="col s12 m9">
    <div class="post-listing">
      <div id="post">
        <h1>Librator, a platform for optimized sequence editing, design, and expression of influenza virus proteins</h1>
        <p>
          Artificial mutagenesis and chimeric/mosaic protein engineering have laid the foundation for antigenic characterization and universal vaccine design for influenza viruses. However, both methods require vast sequence editing and protein expression, which are subject to expensive reagent costs and inefficient workflow, limiting their applicability and the progress of related research. To overcome these shortages, we developed Librator, a system for designing protein sequences of Hemagglutinin (HA) and Neuraminidase (NA) of type A influenza virus and standardized fragments of Gibson Assembly4 cloning. We developed an optimized fragment design based on peptide conservation of all historical HA and NA sequences to ensure fragments are reusable and compatible, allowing for significant reagent savings. Furthermore, with Librator’s graphical user interface and built-in sequence editing functions, biologists can easily mutate target residues and design sequences for antigen probes and chimeric/mosaic proteins efficiently and accurately. Use of Librator in our lab has demonstrated that it can help biologists design HA and NA sequences efficiently and express them economically.
        </p>

        <h2>Updates</h2>
        <div>
          <p><b>Sep,16,2021</b> Librator-v1.0.0-beta version is released. See <a href='https://github.com/WilsonImmunologyLab/Librator/releases/tag/v1.0.0-beta' target="_blank" style="color: green;">here</a> for details.</p>
          <p><b>Apr,29,2021</b> pre-print of Librator has been uploaded to bioRxiv. See <a href='https://www.biorxiv.org/content/10.1101/2021.04.29.441999v1' target="_blank" style="color: green;">here</a> for details.</p>
          <p><b>Apr,18,2021</b> Librator-v1.0.0-alpha version is released. See <a href='https://github.com/WilsonImmunologyLab/Librator/releases/tag/v1.0.0-alpha' target="_blank" style="color: green;">here</a> for details.</p>
        </div>
        <h2>Github Page</h2>
        Librator is developed using Python 3. All the source code are available on <a href="https://github.com/WilsonImmunologyLab/Librator" target="_blank" style="color: green;">Github Page</a>

        <h2>Pre-print</h2>
        For more details about Librator, please refer to our paper: <a href="https://www.biorxiv.org/content/10.1101/2021.04.29.441999v1" target="_blank" style="color: green;">Li, et al., Librator, a platform for optimized sequence editing, design, and expression of influenza virus proteins. 2021</a>

        <h2>System structure of Librator</h2>
        <p>Librator is comprised of a UI layer (GUI), logical layer (all functions) and data layer (SQL databases). Users are allowed to finish all operations using the GUI. All functions can be divided into two broad categories: basic function and advanced function. Basic function includes I/O operations and database (DB) operations, and advanced function includes sequence design/editing, fragment design, phylogenetic analysis and structure visualization.</p>
        <img src="img/FigS4.png" style="margin: auto; width: 1200px;"/>
        
        <h2>Librator enables effciant and accurate design of influenza sequences</h2>
        <p>
          Librator functions as an integrated graphical processing platform for influenza sequences. Librator seamlessly connects nucleotide sequences (from public sequence databases) and lab work (e.g. Gibson cloning, protein expression), and it contains a variety of functions to facilitate biologists’ managing, editing and accessing their sequences, aimed at eventually improving the efficiency of sequence design and expression. This software entrusts all error-prone sequence editing and data processing tasks to the background algorithm, and users are able to design their sequences with a few clicks on the GUI. This feature greatly facilitates the work of biologists who are not familiar with command-line tools, as well as reduces the possibility of artificial mistakes. With the help of Librator, users can complete all sequence design-related operations in this software, and no longer be bothered by either difficult-to-read raw sequences or switching back and forth between different software. 
        </p>
        <img src="img/fig3.png" style="margin: auto; width: 1680px;"/>

        <h2>Fragment design for influenza HA proteins</h2>
        <p>
          Based on a comprehensive investigation of peptide conservation using all historic HA and NA protein sequences available, we determined three (on HA genome) and two (on NA genome) highly conserved regions as joint regions for Gibson clone fragment design. Accordingly, these joint regions split the HA and NA genomes into four and three fragments with appropriate size, respectively, thus balancing the reusability and cost. According to the evolutionary history of influenza HA subtypes, we designed uniform fragments on the basis of a classic H1 sequence (A/California/7/2009, H1N1) and a classic H3 sequence (A/Aichi/2/1968, H3N2). We aligned all group 1 HAs (H1, H2, H5, H6, H8, H9, H11, H12, H13, H16, H17, and H18) to the H1 template and all group 2 HAs (H3, H4, H7, H10, H14, H15) to the H3 template for fragment design. For NAs, we designed uniform fragments for each subtype by aligning each of the NA sequences to the template of their respective subtypes. This template-mapping-based fragment design ensures that all fragments are standardized and not affected by either different batches or sporadic insertion/deletion events (e.g. a deletion 130 between pre-1994 and post-1994 human seasonal H1N1, or insertions in the cleavage site of high pathogenic avian H5 and H7). 
        </p>
        <img src="img/FigureS1.png" style="width: 1680px;"/>
        <img src="https://badges.toozhao.com/badges/01FCTSMC8YEZQAE4KGQ7Q78ZRC/green.svg" />
      </div>
      
    </div>
  </div>
</div>




<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
